docs/specs/consensus/README.md

```diff
@@ -0,0 +1,1124 @@
+            "3",
+            "1",
+            "3",
+            "1",
+            "4",
+            "2",
+            "1",
+            "2",
+            "2",
+            "4",
+            "5"
+          ],
+          "last_vrf_output": "g_1vrXSXLhvn1e4Ap1Ey5e8yh3PFMJT0vZyhZLlTBAA=",
+          "total_currency": "167255800000001000",
+          "curr_global_slot": {
+            "slot_number": "12978",
+            "slots_per_epoch": "7140"
+          },
+          "global_slot_since_genesis": "12978",
+          "staking_epoch_data": {
+            "ledger": {
+              "hash": "jxX6VJ84HaafrKozFRA4qjnni4aPXqXC2H5vQLKSryNpKTXuz1R",
+              "total_currency": "165950000000001000"
+            },
+            "seed": "2vb1Mjvydod6sEwn7qpbejKCfRqugMgyG3MHXXRKcAkwQLRs9fj8",
+            "start_checkpoint": "3NK2tkzqqK5spR2sZ7tujjqPksL45M3UUrcA4WhCkeiPtnugyE2x",
+            "lock_checkpoint": "3NK5G8Xqn1Prh3XoTyZ2tqntJC6X2nVwruv5mEJCL3GaTk7jKUNo",
+            "epoch_length": "1769"
+          },
+          "next_epoch_data": {
+            "ledger": {
+              "hash": "jx7XXjRfJj2mGXmiHQmpm6ZgTxz14udpugyFtw4DefJFpie7apN",
+              "total_currency": "166537000000001000"
+            },
+            "seed": "2vavBR2GfJWvWkpC7yGJQFnts18nHaFjdVEr84r1Y9DQXvnJRhmd",
+            "start_checkpoint": "3NLdAqxtBRYxYbCWMXxGu6j1hGDrpQwGkBDF9QvGxmtpziXQDADu",
+            "lock_checkpoint": "3NL4Eis1pS1yrPdfCbiJcpCCYsHuXY3ZgEzHojPnFWfMK9gKmhZh",
+            "epoch_length": "2084"
+          },
+          "has_ancestor_in_same_checkpoint_window": true,
+          "block_stake_winner": "B62qpBrUYW8SHcKTFWLbHKD7d3FqYFvGRBaWRLQCgsr3V9pwsPSd7Ms",
+          "block_creator": "B62qpBrUYW8SHcKTFWLbHKD7d3FqYFvGRBaWRLQCgsr3V9pwsPSd7Ms",
+          "coinbase_receiver": "B62qpBrUYW8SHcKTFWLbHKD7d3FqYFvGRBaWRLQCgsr3V9pwsPSd7Ms",
+          "supercharge_coinbase": true
+        },
+        "constants": {
+          "k": "290",
+          "slots_per_epoch": "7140",
+          "slots_per_sub_window": "7",
+          "delta": "0",
+          "genesis_state_timestamp": "1609355670000"
+        }
+      }
+    },
+    "protocol_state_proof": "<opaque>",
+    "staged_ledger_diff": "<opaque>",
+    "delta_transition_chain_proof": "<opaque>",
+    "current_protocol_version": "1.1.0",
+    "proposed_protocol_version": "<None>"
+  }
+}
+```
+
+# 5. Algorithms
+
+This section outlines the main algorithms and constructs used by Samasika.
+
+## 5.1 Common
+
+This section outlines some commonly used helpers.
+
+### 5.1.1 `top`
+
+This function returns the last block of a given chain.  The input is a chain `C` and the output is last block of `C` (i.e. the block with greatest height).
+
+```rust
+fn top(C) -> Block
+{
+   return last block of C
+}
+```
+
+### 5.1.2 `cState`
+
+The function returns the consensus state of a block or chain.  The input is a block or chain `X` and the output is the consensus state.
+
+```rust
+fn cState(X) -> Consensus_state
```

Comment:
It's unclear to me why this polymorphism is needed here. You can just call `cState(top(X))` when `X : Chain`.

Reply:
It not needed.  It's only there to keep the high-level pseudocode as simple as possible, i.e. so the reader doesn't need to care whether something is a block or chain...  It is not an implementation recommendation.